File: Core/src/main/java/jokatu/game/input/InputAcceptor.java
package jokatu.game.input;

import jokatu.game.Game;
import jokatu.game.event.GameEvent;
import jokatu.game.player.Player;
import ophelia.collections.BaseCollection;
import ophelia.event.observable.Observable;
import org.jetbrains.annotations.NotNull;

/**
 * Accepts an {@link Input} from a {@link Player}, handles it, and may fire {@link GameEvent}s.  These are to be
 * created in the context of a single {@link Game} and will most likely have access to various parts of the game's
 * state.
 *
 * @param <E> the type of {@link GameEvent} that this may fire
 */
public interface InputAcceptor<E extends GameEvent> extends Observable<E> {

	@NotNull
	BaseCollection<Class<? extends Input>> getAcceptedInputs();

	void accept(@NotNull Input input, @NotNull Player player) throws Exception;
}

File: Core/src/main/kotlin/jokatu/game/input/InputAcceptor.kt
package jokatu.game.input

import jokatu.game.Game
import jokatu.game.event.GameEvent
import jokatu.game.player.Player
import ophelia.event.observable.Observable
import kotlin.reflect.KClass

/**
 * Accepts an [Input] from a [Player], handles it, and may fire [GameEvent]s.  These are to be
 * created in the context of a single [Game] and will most likely have access to various parts of the game's
 * state.
 *
 * @param <E> the type of [GameEvent] that this may fire
 */
interface InputAcceptor<E : GameEvent> : Observable<E> {

	val acceptedInputs: Collection<KClass<out Input>>

	@Throws(Exception::class)
	fun accept(input: Input, player: Player)
}
